wicker/core/datasets.py

```diff
@@ -125,79 +135,133 @@ def __len__(self) -> int:
         pass
 
     @abc.abstractmethod
-    def schema(self) -> DatasetSchema:
-        """Return the schema of the dataset."""
+    def arrow_table(self) -> pyarrow.Table:
+        """Return the pyarrow table with all the metadata fields and pointers of the dataset."""
         pass
 
     @abc.abstractmethod
-    def arrow_table(self) -> pyarrow.Table:
-        """Return the pyarrow table with all the metadata fields and pointers of the dataset."""
+    def schema(self) -> DatasetSchema:
+        """Return the schema of the dataset."""
         pass
 
 
-class S3Dataset(AbstractDataset):
-    """Implementation for a Map-based dataset"""
+class BaseDataset(AbstractDataset):
+    """Provides an implementation for part of the AbstractDataset interface for datasets whose items
+    and schema can be read from AbstractDataStorage concrete class."""
 
```

Comment:
For `BaseDataset`, some of the key arguments like `pa_filesystem`, `path_factory` and `storage` should no longer be `Optional` or take an S3-specific default value. So in the `__init__` function, I just move these arguments higher in the argument list.